Introduce perturbation module (#73)
Qiskit/qiskit-dynamics: docs/userguide/perturbative_solvers.rst

How-to use Dyson and Magnus based solvers
=========================================

> In this tutorial we walk through how to use perturbation-theory based solvers.

Author: Moved the sentence to after the notes. Also added an additional sentence referencing the documentation of the solvers for information about how they work, along with the discussion for background info.

Linking to the class docs is a little bit redundant with the first note, though they are linking to the docs for different things (API documentation v.s. description of what computations the class is actually doing).

> .. note::

Author: Can you elaborate? Not sure what you mean by this.

> .. note::

Author: Changed to a warning - I like the colour difference between warning/note for this.

   strength.
-  :math:`s(t)` is the drive signal, which we will take to be on
   resonance with envelope :math:`f(t) = A \frac{4t (T - t)}{T^2}`
   for a given amplitude :math:`A` and total time :math:`T`.

We will walk through the following steps:

1. Configure ``qiskit-dynamics`` to work with JAX.
2. Construct the model.
3. How-to construct and simulate using the Dyson-based perturbative solver.
4. Simulate using a traditional ODE solver, comparing speed.
5. How-to construct and simulate using the Magnus-based perturbative solver.

1. Configure to use JAX
-----------------------

These simulations will be done with JAX array backend to enable
compilation. See the user guide entry on using JAX for a detailed
breakdown of using JAX with dynamics.

.. jupyter-execute::

    from qiskit_dynamics.array import Array

    # configure jax to use 64 bit mode
    import jax
    jax.config.update("jax_enable_x64", True)

    # tell JAX we are using CPU
    jax.config.update('jax_platform_name', 'cpu')

    # set default backend
    Array.set_default_backend('jax')

2. Construct the model
----------------------

First, construct the model described in the introduction. Here we use a

> higher dimension to observe a difference between the solvers.

Author: Expanded on this point in the doc.

.. jupyter-execute::

    import numpy as np

    dim = 10

    v = 5.
    anharm = -0.33
    r = 0.02

    a = np.diag(np.sqrt(np.arange(1, dim)), 1)
    adag = np.diag(np.sqrt(np.arange(1, dim)), -1)
    N = np.diag(np.arange(dim))

    # static part
    static_hamiltonian = 2 * np.pi * v * N + np.pi * anharm * N * (N - np.eye(dim))
    # drive term
    drive_hamiltonian = 2 * np.pi * r * (a + adag)

    # total simulation time
    T = 1. / r

    # envelope function
    envelope_func = lambda t: t * (T - t) / (T**2 / 4)

3. How-to construct and simulate using the Dyson-based perturbative solver
--------------------------------------------------------------------------

Constructing the Dyson-based perturbative solver requires specifying several
configuration parameters, as well as specifying the structure of the
differential equation more explicitly than using the standard

> :class:`.Solver` object in qiskit-dynamics, which automatically builds

Reviewer: Suggested rewording/ordering
```
Setting up a :class:.`DysonSolver` requires more setup han the standard
:class:`.Solver` as the user must specify several configuration parameters,
along with the structure of the differential equation:

- A :class:`.DysonSolver` requires direct specification of the LMDE to the
  solver. If we are simulating the Schrodinger equation, we need to
  multiply the Hamiltonian terms by ``-1j`` when passing in as LMDE
  operators.
- The :class:`.DysonSolver` is a fixed step solver, with the step size
  being fixed at instantiation. This step size must be chosen in conjunction
  with the ``expansion_order``, to ensure that a suitable accuracy is attained.
- Over each fixed time-step the :class:`.DysonSolver` solves by computing a
   truncated perturbative expansion.
  - To compute the truncated perturbative expansion, the signal envelopes are
    approximated as a linear combination of Chebyshev polynomials.
  - The order of the Chebyshev approximations, along with central carrier frequencies
    for defining the “envelope” of each ``Signal``, must be provided at instantiation.

See the :ref:`API docs <.DysonSolver>` for a more details.
```

Author: Implemented this suggestion with a few edits.

either the Schrodinger or Lindblad equation based on the inputs.

See the API docs for :class:`.DysonSolver` for a more detailed
explanation, but some general comments on its instantiation and usage:

- :class:`.DysonSolver` requires direct specification of the LMDE to the
  solver. As we are simulating the Schrodinger equation, we need to
  multiply the Hamiltonian terms by ``-1j``.
- :class:`.DysonSolver` is a fixed step solver, with the step size
  being fixed at instantiation. This step size must be chosen in conjunction
  with the ``expansion_order``, to ensure that a suitable accuracy is attained.
- Over each fixed time-step:

  - :class:`.DysonSolver` solves by computing a truncated perturbative
    expansion.
  - To compute the truncated perturbative expansion, the signal envelopes are
    approximated as a linear combination of Chebyshev polynomials.
  - The order of the Chebyshev approximations, along with central carrier frequencies
    for defining the “envelope” of each ``Signal``, must be provided at instantiation.

> .. jupyter-execute::

Reviewer: If going with above ordering, can add a leading sentance to next block

```suggestion
For our example Hamiltonian we configure the :class:`.DysonSolver` as follows:

.. jupyter-execute::
```

Author: Added this sentence.

    %%time

    from qiskit_dynamics import DysonSolver

    dt = 0.1
    dyson_solver = DysonSolver(
        operators=[-1j * drive_hamiltonian],
        rotating_frame=-1j * static_hamiltonian,
        dt=dt,
        carrier_freqs=[v],
        chebyshev_orders=[1],
        expansion_order=7,
        integration_method='jax_odeint',
        atol=1e-12,
        rtol=1e-12
    )

> 

Reviewer: Can you add a sentence explaining the chosen values for expansion_oder and chebyshev orders?

Author: Add a few sentences giving some context to the parameters.

> Construct a function that simulates the system for the pulse sequence

Reviewer: Things get confusing here. I think there should be longer paragraph explaining in a little more detail how to use the dyson solver to simulate something.

eg something like: 
> To use the Dyson solver we wrap it in a function which evaluates the perturbation expansion over a fixed length parameterized drive envelope and returns the final system states. This can then be used to compare the accuracy and performance with the regular solver over the same drive pulse. We also use JAX to just-in-time compile this function to improve performance for repeated evaluations for different amplitude parameter values.

with a given amplitude.

.. jupyter-execute::

>     from qiskit_dynamics import Signal

Reviewer: All imports at the top
```suggestion
    from qiskit_dynamics import Signal
    from jax import jit
```

Author: Moved the JAX import.

>     def dyson_sim(amp):

Reviewer: Add doc string

```suggestion
    def dyson_sim(amp):
        """Evaluate dyson solver over a fixed drive envelope with given amplitude"""
```

        drive_signal = Signal(lambda t: Array(amp) * envelope_func(t), carrier_freq=v)
        return dyson_solver.solve(
            signals=[drive_signal],
            y0=np.eye(dim, dtype=complex),
            t0=0.,
            n_steps=int(T // dt)
        ).y[-1]

    from jax import jit

>     jit_dyson_sim = jit(dyson_sim)

Reviewer: ```suggestion
    # Jit the evaluation function to improve performance for repeated calls
    jit_dyson_sim = jit(dyson_sim)
```

Author: Added.

> First run includes compile time.

Reviewer: Needs some explanantion
```suggestion
Because we are using a just-in-time compiled function, the first call will take longer as it compiles the functions. We can time this compilation using the Juypter magic method:
```

Author: I think this might be getting too verbose/explain-y for a how-to guide. I know this one has to be more verbose/explain-y than the others when it comes to details/usage of the solvers, but I don't think we need to explain details of JAX compilation/magic methods.

I've added an explicit reference to the userguide entry on using JAX in the "Configure to use JAX" section of this doc. Let me know if this alleviates your concerns.

.. jupyter-execute::

>     %time yf_dyson = jit_dyson_sim(1.).block_until_ready()

Reviewer: Maybe add a note on `block_until_ready` to previous sentence for people not familar with jax?
> Note that the `.block_until_ready` is required for timing the jax evaluation since the returned jax array is evaluated lazily.

Author: Same as above.

> Second run demonstrates the speed of the solver.

Reviewer: ```suggestion
Once JIT compilation has been performance we can benchmark the performance of the jit-compiled solver:
```

            t_span=[0., int(T // dt) * dt],
            y0=np.eye(dim, dtype=complex),
            signals=[drive_signal],
            method='jax_odeint',
            atol=tol,
            rtol=tol
        )
        return res.y[-1]

Simulate with low tolerance for comparison to high accuracy solution.

.. jupyter-execute::

    yf_low_tol = ode_sim(1., 1e-13)
    np.linalg.norm(yf_low_tol - yf_dyson)

For speed comparison, compile at a tolerance with similar accuracy.

.. jupyter-execute::

    jit_ode_sim = jit(lambda amp: ode_sim(amp, 1e-8))

    %time yf_ode = jit_ode_sim(1.).block_until_ready()

Measure compiled time.

.. jupyter-execute::

    %time yf_ode = jit_ode_sim(1.).block_until_ready()

Confirm similar accuracy solution.

.. jupyter-execute::

    np.linalg.norm(yf_low_tol - yf_ode)

Here we see that, once compiled, the Dyson-based solver has a
significant speed advantage over the traditional solver, at the expense
of the initial compilation time and the technical aspect of using the solver.

5. How-to construct and simulate using the Magnus-based perturbation solver
---------------------------------------------------------------------------

> Next, build the Magnus-based perturbative solver. The :class:`.MagnusSolver` uses the

Reviewer: ```suggestion
We repeat our example using the Magnus-based perturbative solver. Setup of the :class:`.MagnusSolver`
is the same to the :class:`.DysonSolver`, but uses the Magnus expansion and
matrix exponentiation to simulate over each fixed time step.
```
Im not sure if it should be *the same as * or *similar to*?

Author: Put this in - I think "similar to" is good here.
